opds2_import.py

```diff
@@ -0,0 +1,830 @@
+            reversed(
+                sorted(
+                    publication.images.links, key=lambda link: (link.width, link.height)
+                )
+            )
+        )
+        image_links = []
+
+        if len(sorted_raw_image_links) > 0:
+            cover_link = self._extract_link(
+                sorted_raw_image_links[0],
+                feed_self_url,
+                default_link_rel=Hyperlink.IMAGE,
+            )
+            image_links.append(cover_link)
+
+        if len(sorted_raw_image_links) > 1:
+            cover_link = self._extract_link(
+                sorted_raw_image_links[1],
+                feed_self_url,
+                default_link_rel=Hyperlink.THUMBNAIL_IMAGE,
+            )
+            image_links.append(cover_link)
+
+        self._logger.debug(
+            u"Finished extracting image links from {0}: {1}".format(
+                encode(publication.images), encode(image_links)
+            )
+        )
+
+        return image_links
+
+    def _extract_links(self, publication, feed_self_url):
+        """Extract a list of LinkData objects from a list of webpub-manifest-parser links.
+
+        :param publication: Publication object
+        :type publication: ast_core.Publication
+
+        :param feed_self_url: Feed's self URL
+        :type feed_self_url: str
+
+        :return: List of links metadata
+        :rtype: List[LinkData]
+        """
+        self._logger.debug(
+            u"Started extracting links from {0}".format(encode(publication.links))
+        )
+
+        links = []
+
+        for link in publication.links:
+            link_metadata = self._extract_link(link, feed_self_url)
+            links.append(link_metadata)
+
+        description_link = self._extract_description_link(publication)
+        if description_link:
+            links.append(description_link)
+
+        image_links = self._extract_image_links(publication, feed_self_url)
+        if image_links:
+            links.extend(image_links)
+
+        self._logger.debug(
+            u"Finished extracting links from {0}: {1}".format(
+                encode(publication.links), encode(links)
+            )
+        )
+
+        return links
+
+    def _extract_media_types_and_drm_scheme_from_link(self, link):
+        """Extract information about content's media type and used DRM schema from the link.
+
+        :param link: Link object
+        :type link: ast_core.Link
+
+        :return: 2-tuple containing information about the content's media type and its DRM schema
+        :rtype: List[Tuple[str, str]]
+        """
+        self._logger.debug(
+            u"Started extracting media types and a DRM scheme from {0}".format(
+                encode(link)
+            )
+        )
+
+        media_types_and_drm_scheme = []
+
+        if link.properties:
+            if (
+                link.properties.availability.state
+                == opds2_ast.OPDS2AvailabilityType.AVAILABLE.value
+            ):
```

Comment:
I see this in your example feed in python-webpub-manifest parser but I don't think this is part of OPDS 2 or RWPM. What does this mean, and why do you ignore format information if this information is missing or its value is other than 'available'?

Reply:
It's [a part](https://github.com/opds-community/drafts/blob/master/schema/properties.schema.json#L80) of the OPDS 2.0 schema. I found a lot of examples of using `availability` property in the Pret Numerique feed.

For example, the example below show that the book is available:
```
{
      "type": "application/opds-publication+json",
      "rel": "http://opds-spec.org/acquisition/borrow",
      "href": "http://montreal.pretnumerique.ca/v1/media/feedhttps-www-feedbooks-com-item-3123893-9782811221126-epub/loans_activation.opds2?_resource_id=5d4da62423579414ddd09c68",
      "properties": {
        "availability": {
          "state": "available"
        },
        "indirectAcquisition": [{
          "type": "application/vnd.adobe.adept+xml",
          "child": [{
            "type": "application/epub+zip"
          }]
        }]
      }
    }
```

The other example below shows that the books is not available until `2020-10-23T17:03:36.178Z`:
```{
      "type": "application/opds-publication+json",
      "rel": "http://opds-spec.org/acquisition/borrow",
      "href": "http://montreal.pretnumerique.ca/v1/bookings.opds2?_resource_id=5f775a842357946dbe59f5af\u0026medium_id=ENQC109242-9782897059637-pdf",
      "properties": {
        "availability": {
          "state": "unavailable",
          "until": "2020-10-23T17:03:36.178Z"
        },
        "indirectAcquisition": [{
          "type": "application/vnd.adobe.adept+xml",
          "child": [{
            "type": "application/pdf"
          }]
        }, {
          "type": "application/vnd.readium.lcp.license.v1.0+json",
          "child": [{
            "type": "application/pdf"
          }]
        }]
      }
    }
```

Since `availability` property is optional, I also add another condition allowing to add formats of acquisition links not having this property.

Reply:
I see, it's in the schema but not written up in the spec yet. I created these options for OPDS 1.2 but had to put it in an extension so I'm glad it'll be in the core spec for 2.0.

Reply:
Updated the code accordingly